crates/dht/src/time.rs

```diff
@@ -0,0 +1,1051 @@
+//! amount of time left over that cannot be partitioned into smaller slices.
+//!
+//! Because recent time is expected to change more frequently, combined hashes for partial slices
+//! are stored in memory. Full slices are stored in the Kitsune2 op store.
+//!
+//! The algorithm for partitioning time is as follows:
+//!   - Reserve a minimum amount of recent time as a sum of possible partial slice sizes.
+//!   - Allocate as many full slices as possible.
+//!   - Partition the remaining time into smaller slices. If there is space for two slices of a
+//!     given size, then allocate two slices of that size. Otherwise, allocate one slice of that
+//!     size. Larger slices are allocated before smaller slices.
+//!
+//! As time progresses, the partitioning needs to be updated. This is done by calling the
+//! [PartitionedTime::update] method. The update method will:
+//!    - Run the partitioning algorithm to determine whether new full slices can be allocated and
+//!      how to partition the remaining time into partial slices.
+//!    - Store the combined hash of any new full slices in the Kitsune2 op store.
+//!    - Store the combined hash of any new partial slices in memory.
+//!    - Update the `next_update_at` field to the next time an update is required.
+//!
+//! As an example, consider a factor of 9. This means that full slices are 2^9 = 512 times the [UNIT_TIME].
+//! With a unit time of 15 minutes, that means full slices are 512 * 15 minutes = 7680 minutes = 128 hours.
+//! So every 5 days, a new full slice is created.
+//! The partial slices reserve at least 2^8 + 2^7 ... 2^0 = 519 times the [UNIT_TIME], so 519 * 15 minutes
+//! = 7785 minutes = 129.75 hours. That gives roughly 5 days of recent time.
+//!
+//! A lower factor allocates less recent time and more slices to be stored but is more granular when
+//! comparing time slices with another peer. A higher factor allocates more recent time and fewer
+//! slices to be stored but is less granular when comparing time slices with another peer.
+
+use crate::constant::UNIT_TIME;
+use kitsune2_api::{DynOpStore, K2Error, K2Result, OpId, Timestamp};
+use std::time::Duration;
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(Clone, PartialEq))]
+pub struct PartitionedTime {
+    /// The timestamp at which the time partitioning starts.
+    ///
+    /// This must be a constant among peers who need to compute the same time slices.
+    origin_timestamp: Timestamp,
+    /// The factor used to determine the size of the time slices.
+    ///
+    /// The size of a time slice is 2^factor * [UNIT_TIME].
+    /// Partial slices are created for "recent time" and are created in decreasing sizes from
+    /// 2^factor * [UNIT_TIME] down to 2^0 * [UNIT_TIME].
+    factor: u8,
+    /// The number of full time slices that have been stored.
+    ///
+    /// These full slices are always of size 2^factor * [UNIT_TIME].
+    full_slices: u64,
```

Comment:
It is, the full slices have their hashes stored on the host. They will change much less frequently and ideally we wouldn't permit them to change at all after some amount of time has passed

Reply:
I understand, but since we're comparing them every time the number of slices matches the remote's number of slices, it would be more resource-efficient to keep them in memory than restoring them from a database each time. They shouldn't take up much space in memory.

Reply:
I'm thinking that in the majority of cases, we don't need to retrieve them. If we have a full slice time of something like 5 days then over time we end up with quite a few hashes to keep in memory. Not an unreasonable number but when it's required across a hash partition too, that's a fair bit. 

In the majority of gossip rounds, where both nodes have been online since the last gossip round, all we really need to check is "do our all our previous 5 day slices, combined, still agree". We can use a count of full slices to understand which side needs to recalculate if they are at a different point (i.e. one has just updated their ParititionedTime and the other hasn't).

So we probably end up storing the current and maybe 1 previous combined hash of all full slices here. Yes, we will end up adding something here, but I intentionally left out storing them all. It's only when we find a historical mismatch that we need to grab more historical slices. But even then, we only need to fetch from where the remote is telling us they match up to.

Reply:
I hope this makes some sense, it's a little rambly! :)

Reply:
Gotcha, yes. I was assuming that we don't store a timestamp when two peers last spoke. If they do, even just in memory, then you only have to restore those full slice hashes once and combine them to a... combined hash. That combined hash may be added, understood. That's sound.